File: solutions/_1046_last_stone_weight/solution.py
from typing import List
from heapq import heappop, heappush, heapify


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones = [-s for s in stones]
        heapify(stones)

        while len(stones) > 1:
            y = heappop(stones)
            x = heappop(stones)
            if x != y:
                heappush(stones, y - x)

        stones.append(0)
        return abs(stones[0])


# https://leetcode.com/problems/last-stone-weight/

File: solutions/_1046_last_stone_weight/solution.js
const { MaxPriorityQueue } = require("@datastructures-js/priority-queue");

/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function (stones) {
  const pq = new MaxPriorityQueue();
  for (const stone of stones) {
    pq.enqueue(stone);
  }

  while (pq.size() > 1) {
    const y = pq.dequeue();
    const x = pq.dequeue();
    if (y != x) {
      pq.enqueue(y - x);
    }
  }

  return pq.isEmpty() ? 0 : pq.front();
};

module.exports = { lastStoneWeight };

/*
https://leetcode.com/problems/last-stone-weight/
*/
